plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java

```diff
@@ -125,6 +128,124 @@ public void respond(
+      Document document,
+      final Map<Annotation, org.eclipse.che.ide.api.editor.text.Position> annotations,
+      final boolean goToClosest,
+      List<Diagnostic> resultingDiagnostics) {
+    int invocationLocation = range.getStartOffset();
+    if (goToClosest) {
+      LinearRange line =
+          document.getLinearRangeForLine(
+              document.getPositionFromIndex(range.getStartOffset()).getLine());
+      int rangeStart = line.getStartOffset();
+      int rangeEnd = rangeStart + line.getLength();
+
+      ArrayList<org.eclipse.che.ide.api.editor.text.Position> allPositions = new ArrayList<>();
+      List<DiagnosticAnnotation> allAnnotations = new ArrayList<>();
+      int bestOffset = Integer.MAX_VALUE;
+      for (Annotation a : annotations.keySet()) {
+        org.eclipse.che.ide.api.editor.text.Position pos = annotations.get(a);
+        if (pos != null && isInside(pos.offset, rangeStart, rangeEnd)) { // inside our range?
+          allAnnotations.add((DiagnosticAnnotation) a);
+          allPositions.add(pos);
+          bestOffset = processAnnotation(a, pos, invocationLocation, bestOffset);
+        }
+      }
+      if (bestOffset == Integer.MAX_VALUE) {
+        return invocationLocation;
+      }
+      for (int i = 0; i < allPositions.size(); i++) {
+        org.eclipse.che.ide.api.editor.text.Position pos = allPositions.get(i);
+        if (isInside(bestOffset, pos.offset, pos.offset + pos.length)) {
+          resultingDiagnostics.add(allAnnotations.get(i).getDiagnostic());
+        }
+      }
+
+      return bestOffset;
+    } else {
+      // iteration with range never returns anything; need to filter ourselves.
+      // https://github.com/eclipse/che/issues/4338
+      resultingDiagnostics.addAll(
+          annotations
+              .entrySet()
+              .stream()
+              .filter((e) -> e.getValue().overlapsWith(range.getStartOffset(), range.getLength()))
+              .map(Entry::getKey)
+              .map(a -> (DiagnosticAnnotation) a)
+              .map(DiagnosticAnnotation::getDiagnostic)
+              .collect(Collectors.toList()));
+    }
+    return invocationLocation;
+  }
+
+  /**
+   * Tells is the offset is inside the (inclusive) range defined by start-end.
+   *
+   * @param offset the offset
+   * @param start the start of the range
+   * @param end the end of the range
+   * @return true iff offset is inside
+   */
+  private boolean isInside(int offset, int start, int end) {
+    return offset == start
+        || offset == end
+        || (offset > start && offset < end); // make sure to handle 0-length ranges
+  }
+
+  private int processAnnotation(
+      Annotation annot,
+      org.eclipse.che.ide.api.editor.text.Position pos,
+      int invocationLocation,
+      int bestOffset) {
+    final int posBegin = pos.offset;
+    final int posEnd = posBegin + pos.length;
+    if (isInside(invocationLocation, posBegin, posEnd)) { // covers invocation location?
```

Comment:
Now I see, you're using `isInside` method all over the pull request and give an explanation in comments to make it clearer for the reader. That is very nice of you however I would still recommend to avoid using comments in such cases and make the code itself more declarative instead.

Reply:
Again. Already commented on the source of these comments.